### PR TITLE
Codegen: Do less work in dry-runs for sharded files

### PR DESCRIPTION
**Pull request by peterbell10**
This improves a dry-run of `gen.py` from 0.80s to 0.45s.

`FileManager` in `dry_run` mode doesn't actually need to compute the
environment; it just records the filenames that would have been
written.

cc @ezyang @bhosmer @bdhirsh

### CODE REVIEW

File: tools/codegen/utils.py
Comment (albanD) [inline tools/codegen/utils.py:204]: Should we skip generating this altogether?

Comment (peterbell10) [reply tools/codegen/utils.py:204]: We still need to call `write_with_template` to update the `filenames` list.

Comment (albanD) [reply tools/codegen/utils.py:204]: Ho what I meant is that the caller doesn't need to provide a valid items and so could skip generating it?
I know the loading of the yaml is especially expensive, so if we can skip that, it would be a nice speedup.
But this is quite fast anyways so maybe not worth it?

Comment (peterbell10) [reply tools/codegen/utils.py:204]: Oh I see. Once the `--per-operator-headers` flag lands then we really do need to read `native_functions.yaml` even for dry runs.

Comment (albanD) [reply tools/codegen/utils.py:204]: Ok!
I'm working on the import of that other PR.
Let's delay this one until then then?

Comment (peterbell10) [reply tools/codegen/utils.py:204]: Sure you can delay this, it's a fairly small gain in the grand scheme of things. To be clear though, this PR doesn't depend on the other stack.

**Comment by peterbell10** [comment]
@albanD ping

**Comment by albanD** [comment]
Can you rebase this on top of master for the CI to run again? Should be good to go after.

**Review by albanD — APPROVED** [review]

**Comment by peterbell10** [comment]
@alband rebased and tests are passing

**Pull request closed without merging**
